File: sdks/kotlin/src/main/kotlin/dev/orion/sdk/model/ModelRef.kt
package dev.orion.sdk.model

/** Identifies a model within a registered provider namespace. */
internal data class ModelRef(
    /** Stable provider key resolved by `ModelRegistry`. */
    public val provider: String,
    /** Provider-specific model identifier. */
    public val model: String,
) {

    init {
        require(provider.isNotBlank()) { "model provider must not be blank" }
        require(model.isNotBlank()) { "model identifier must not be blank" }
    }

    /** Factories for model references. */
    public companion object {

        /**
         * Parses `provider:model` notation.
         *
         * @throws IllegalArgumentException when either component is missing.
         */
        public fun parse(value: String): ModelRef {

            val separator = value.indexOf(':')

            require(separator > 0 && separator < value.lastIndex) {
                "model reference must use provider:model notation"
            }

            return ModelRef(value.substring(0, separator), value.substring(separator + 1))

        }

    }

}

File: sdks/javascript/src/model/model-ref.ts
/** Serializable provider and model identity used at protocol boundaries. */
export interface ModelRefValue {

  /** Stable provider namespace. */
  readonly provider: string;

  /** Provider-specific model identifier. */
  readonly model: string;

}

/** Selects a model within a stable provider namespace. */
export class ModelRef {

  /** Stable provider key. */
  readonly provider: string;

  /** Provider-specific model identifier. */
  readonly model: string;

  /**
   * Creates a model reference from explicit components.
   *
   * @throws TypeError when either component is empty.
   */
  constructor(provider: string, model: string) {

    if (!provider.trim() || !model.trim()) {
      throw new TypeError("model provider and identifier must be non-empty");
    }

    this.provider = provider;
    this.model = model;

  }

  /**
   * Parses `provider:model` notation.
   *
   * @throws TypeError when the notation omits either component.
   */
  static parse(value: string): ModelRef {

    const index = value.indexOf(":");
    if (index <= 0 || index === value.length - 1) {
      throw new TypeError("model reference must use provider:model notation");
    }

    return new ModelRef(value.slice(0, index), value.slice(index + 1));

  }

}
